doc/design.md

```diff
@@ -0,0 +1,1268 @@
+  For the time being it can be supported by simply holding it in a zero-dimensional variable.
+
+##### HistogramData::Histogram
+
+`Dataset` will store histogram data as a single multi-dimensional array, i.e., *not* as a variable holding histograms.
+However, we may still want to operate on a single histogram at a time in some cases.
+A dataset with one-dimensional data variables `Data::Value` and `Data::Variance` alongside a one-dimensional coordinate variable such as `Coord::Tof` is a histogram.
+- Arithmetic operations as well as many of the other basic operations support most of what would be required from a histogram type.
+
+##### DataObjects::EventWorkspace
+
+In contrast to histograms in `DataObjects::Workspace2D`, the event lists stored in `DataObject::EventWorkspace` have varying length.
+Therefore, the events cannot be stored as a single multi-dimensional array.
+- A one-dimensional variable `Data::Events` with `Dim::Spectrum` holds the lists of the events.
+- The lists of events is a type similar to `DataObjects::EventList`.
+  One option is to use `Dataset` for this, that is the dataset would contain many datasets, one for each event list.
+
+See also the [working example](../test/EventWorkspace_test.cpp).
+
+##### DataObjects::EventList
+
+<img src="dataset-EventList.png" height="30%" width="30%">
+
+A one-dimensional dataset without coordinate variables can be used to represent an event list.
+- `Dim::Event` is used as the only dimension for all variables.
+- `Data::Tof` and `Data::PulseTime` make up the key part of data.
+- Extra variables can be added to support weighted events.
+- `concatenate` provides a merge operation.
+- `sort` provides sorting by time-of-flight or pulse-time.
+- `filter` provides filtering by time-of-flight or pulse-time.
+
+
+### <a name="overview-python-exports"></a>Python exports
+
+All relevant high-level classes and functions are fully exposed via a Python interface.
+- The C++ API is modern and is thus easily mapped to a "pythonic" dataset library.
+- `Dataset` itself is very similar to Python `dict`.
+- Data of variables containing plain old data such as floating point values is exposed as readable and writable `numpy.ndarray`.
+  These are true multi-dimensional arrays, i.e., in contrast to the current `API::MatrixWorkspace::readX()` and `API::MatrixWorkspace::dataX()` it is *not* a separate array for every spectrum.
+
+
+## <a name="design-components"></a>Design components
+
+The following is only an incomplete description.
+For more details and examples the prototype code serves as a reference.
+A good starting point that gives an overview of how `Dataset` may be *used* is given by [`TableWorkspace` example](../test/TableWorkspace_test.cpp), [`Workspace2D` example](../test/Workspace2D_test.cpp), and [`EventWorkspace` example](../test/EventWorkspace_test.cpp).
+
+
+### <a name="design-components-dataset"></a>`class Dataset`
+
+`class Dataset` is literally just a set or list of variables.
```

Comment:
I wonder if the intended public APIs of these classes can be explicitly listed in the document with their short descriptions? The class descriptions are clear though 👍

Reply:
Can do so once we moved into detailed design, for now many of the details are still changing.